Generator crashes if a referenced project declares an internal SupportedOSPlatformAttribute

If a referenced project declares `System.Runtime.Versioning.SupportedOSPlatformAttribute` internally, CsWin32 crashes.

(This was an attempt to create a repro for a different situation in which a referenced *binary* with an internal definition of `System.Runtime.Versioning.SupportedOSPlatformAttribute` causes CsWin32 to think it doesn't need to emit a declaration for SupportedOSPlatformAttribute. It then emits usages of that attribute, causing CS1022 errors in generated code because the type is not accessible.)

I believe this is probably due to use of `Compilation.GetTypeByMetadataName` which has [problems](https://github.com/dotnet/roslyn/issues/52399).

## Repro steps

Extract and run `dotnet build`: [Repro.zip](https://github.com/microsoft/CsWin32/files/7003448/Repro.zip)

> CSC : warning CS8785: Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'InvalidOperationException' with message 'Sequence contains no elements'

### Context

- CsWin32 version: `0.1.506-beta`, 
- Target Framework: `net48`

--- a/src/Microsoft.Windows.CsWin32/Generator.cs
+++ b/src/Microsoft.Windows.CsWin32/Generator.cs
@@ -303,7 +303,8 @@ namespace Microsoft.Windows.CsWin32
 
             this.canCallCreateSpan = this.compilation?.GetTypeByMetadataName(typeof(MemoryMarshal).FullName)?.GetMembers("CreateSpan").Any() is true;
             this.generateDefaultDllImportSearchPathsAttribute = this.compilation?.GetTypeByMetadataName(typeof(DefaultDllImportSearchPathsAttribute).FullName) is object;
-            if (this.compilation?.GetTypeByMetadataName("System.Runtime.Versioning.SupportedOSPlatformAttribute") is { } attribute)
+            if (this.compilation?.GetTypeByMetadataName("System.Runtime.Versioning.SupportedOSPlatformAttribute") is { } attribute
+                && (attribute.DeclaredAccessibility == Accessibility.Public || attribute.ContainingAssembly.GivesAccessTo(this.compilation.Assembly)))
             {
                 this.generateSupportedOSPlatformAttributes = true;
                 AttributeData usageAttribute = attribute.GetAttributes().Single(att => att.AttributeClass?.Name == nameof(AttributeUsageAttribute));

